File: 234-palindrome-linked-list/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* EndOfFirstHalf(ListNode* head){
        int count = 0;
        ListNode* temp = head;

        while(temp->next){
            temp = temp->next;
            count++;
        }

        count /= 2;
        temp = head;
        while(count--){
            head = head->next;
        }

        return temp;
    }

    ListNode* StartOfReverse(ListNode* root){
        ListNode* prev = NULL;
        ListNode* curr = root;

        while(root->next){
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        curr->next = prev;

        return prev;
    }

    bool isPalindrome(ListNode* head) {
        stack<int> st;
        int count = 0;
        ListNode* temp = head;

        while(temp){
            st.push(temp->val);
            count++;
            temp = temp->next;
        }

        count /= 2;
        temp = head;

        while(count--){
            if(st.top() != temp->val) return false;
            st.pop();
            temp = temp->next;
        }

        return true;
    }
};

File: 234-palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        n = 0
        curr = head
        while curr:
            curr = curr.next
            n += 1
        
        mid = n//2

        curr = head
        while mid > 0:
            curr = curr.next
            mid -= 1

        nextCurr = curr.next
        
        while nextCurr:
            temp = nextCurr.next
            nextCurr.next = curr
            curr = nextCurr
            nextCurr = temp

        mid = n//2
        while curr is not head and mid > 0:
            if curr.val is not head.val:
                return False
            curr = curr.next
            head = head.next
            mid -= 1
            
        return True
        